src/transactions.rs

```diff
@@ -367,18 +400,31 @@ impl<L: Ledger + Serialize + DeserializeOwned> Transactions<L> {
     }
 
     /// Remove a transaction from the pending index when it is known to have timed out
-    pub async fn remove_expired(&mut self, timeout: u64) -> Result<(), KeystoreError<L>> {
+    pub fn remove_expired(
+        &mut self,
+        timeout: u64,
+    ) -> Result<Vec<Transaction<L>>, KeystoreError<L>> {
+        let mut removed = vec![];
         if let Some(expiring_uids) = self.expiring_txns.index().get(&timeout).cloned() {
             for uid in expiring_uids.iter() {
                 let editor = self.get_mut(uid)?;
-                editor
-                    .set_status(TransactionStatus::Rejected)
-                    .clear_timeout()
-                    .save()?;
+                // Only return transactions that are not completed because other
+                // transactions did not actually expire.  We still remove all transactions
+                // from they index becuase we'll never lookup for this timeout again.
+                // Also update the status of the expired transactions
+                if editor.transaction.status() == TransactionStatus::Pending {
+                    removed.push(editor.transaction.clone());
+                    editor
+                        .set_status(TransactionStatus::Rejected)
```

Comment:
We could actually have a `TimedOut` status now. Before we couldn't because querying the status on demand, we couldn't differentiate between a timed out transaction and an invalid one. Something to consider later on, maybe.